secrethandshake/state.go

```diff
@@ -173,32 +180,20 @@ func (s *State) verifyClientAuth(data []byte) bool {
 
 	s.hello = make([]byte, 0, len(data)-16)
 
-	var nonce [24]byte // always 0?
-	var openOk bool
-	s.hello, openOk = box.OpenAfterPrecomputation(s.hello, data, &nonce, &s.secret2)
+	var (
+		nonce  [24]byte // always 0?
+		openOk bool
+		sig    [ed25519.SignatureSize]byte
+		public [ed25519.PublicKeySize]byte
+	)
 
-	var sig [ed25519.SignatureSize]byte
-	var public [ed25519.PublicKeySize]byte
-	/* TODO: is this const time!?!
-
-	   this is definetly not:
-	   if !openOK {
-	   	s.hello = nullHello
-	   }
-	   copy(sig, ...)
-	   copy(pub, ...)
-	*/
-	if openOk {
-		copy(sig[:], s.hello[:ed25519.SignatureSize])
-		copy(public[:], s.hello[ed25519.SignatureSize:])
-	} else {
-		copy(sig[:], nullHello[:ed25519.SignatureSize])
-		copy(public[:], nullHello[ed25519.SignatureSize:])
-	}
+	s.hello, openOk = box.OpenAfterPrecomputation(s.hello, data, &nonce, &s.secret2)
+	// subtle API requires an int containing 0 or 1, we only have bool.
+	// we can't branch because openOk is secret.
+	okInt := int(*((*byte)(unsafe.Pointer(&openOk))))
 
-	if lo25519.IsEdLowOrder(sig[:32]) {
-		openOk = false
-	}
+	subtle.ConstantTimeCopy(okInt, sig[:], s.hello[:ed25519.SignatureSize])
```

Comment:
This looks much nicer but it doesn't work. The reason why I added the `nullHello`s is that you can't do the slicing if `openOk` is false because `s.hello` is nil, too!

I added a test case to this branch.